### DISCUSSION TITLE
Fallback page works in dev but errors in production

### DISCUSSION THREAD

**Discussion by brandonpittman** [vercel/next.js#13023] [Help]
I’ve got a fallback page that reads the file system for Markdown in getStaticProps. The fallback page works in dev, but in production, it shows the loading state while router.isFallback is true and then the server errors out. Here’s a typical error. 

```
[GET] /_next/data/ZrNsHZ4IschKCZ7AEV5cr/tags/mac-os.json
22:45:31:14
2020-05-17T13:45:31.218Z	3f2bd3f5-e11d-4373-81b3-bf8779acb3c9	ERROR	Error: ENOENT: no such file or directory, scandir ‘/var/task/_posts’
    at Object.readdirSync (fs.js:872:3)
    at getPostSlugs (/var/task/.next/serverless/pages/tags/[slug].js:3590:52)
    at getAllPosts (/var/task/.next/serverless/pages/tags/[slug].js:3632:17)
    at getStaticProps (/var/task/.next/serverless/pages/tags/[slug].js:129:85)
    at renderToHTML (/var/task/node_modules/next/dist/next-server/server/render.js:277:30)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async renderReqToHTML (/var/task/.next/serverless/pages/tags/[slug].js:2919:22)
    at async Server.render (/var/task/.next/serverless/pages/tags/[slug].js:2966:22) {
  errno: -2,
  syscall: ‘scandir’,
  path: ‘/var/task/_posts’
}
```

Am I using this feature wrong?

**Comment by rafinskipg** [comment]
When you say "works in dev" you mean it works in your computer? or that it works on a "dev" environment that is in vercel?

Let me share some code on how I read the static markdown files

Giving I have a page called `[static-page].js`

```javascript

import React from 'react'
import fs from 'fs'
import marked from 'marked'
import matter from 'gray-matter'
import path from 'path'

const Page = ({ htmlString, data }) => {
  return (
    
     // Here your component with dangerouslySetInnerHtml  to print the markdown
  )
}

export const getStaticPaths = async () => {
  const files = fs.readdirSync('static-pages') // static pages in on the root folder

  const paths = files.map((filename) => ({
    params: {
      ['static-page']: filename.replace('.md', ''),
    },
  }))

  return {
    paths,
    fallback: false,
  }
}

export const getStaticProps = async ({ params }) => {
  const slug = params['static-page']
  const markdownWithMetadata = fs
    .readFileSync(path.join('static-pages', slug + '.md'))
    .toString()
  const parsedMarkdown = matter(markdownWithMetadata)

  return {
    props: {
      data: parsedMarkdown.data,
      htmlString: marked(parsedMarkdown.content),
    },
  }
}

export default Page


```

**Comment by brandonpittman** [comment]
@rafinskipg And this works with `fallback: true`? Cause that's where my issue is  popping up.

**Reply by rafinskipg** [reply]
sorry I didn't try it with fallback. I misunderstood the question

**Reply by brandonpittman** [reply]
@rafinskipg It’s okay. I should have clearly stated in the title that it was about fallback pages.

**Comment by dotlouis** [comment]
I got bitten by a similar issue. It turned out I was misunderstanding how environment variables get loaded in the different getStaticProps() contexts.
Here is my answer from your [other thread](https://github.com/vercel/next.js/discussions/13013#discussioncomment-25484)

> Does your `getStaticProps` use an environment variable?
> I stumbled a similar misunderstanding. It may be unrelated, but posting here just in case.
> 
> What's not immediately clear about this feature is that `getStaticProps()` runs multiple times.
> Once during the _build_ context (when you run next build), but also every time you request a new page that matches the route (SSG), to potentially build a page or fall back to whatever you chose to display as a fallback.
> 
> The thing is that, during the normal build phase. It can read environment variables from the **build context**. But in the SSG phase, it _can't_ and needs to read them from the **runtime context**.
> If you set these variables via the Vercel dashboard, it should work, as there are made available in both contexts.
> However, if like me, you have to set them via the [deployment API](https://vercel.com/docs/api#endpoints/deployments/create-a-new-deployment), or [`vercel.json`](https://vercel.com/docs/configuration#project/env), you have to set them in both `env` AND `build.env` for getStaticProps() to work in SSG.
> 
> I do think that with the lines blurring between build-time and run-time, the Vercel team is trying to simplify this whole environment variables thing. They just need to update the API subsequently ;)
> 
> Hope it helps
